File: platform/playstation/harness/playstation_turn_probe.lua
-- The host half of the turn check: what the emulator says is on screen, at
-- every instant the executable settles.
--
-- `playstation_probe.lua` photographs one still picture at the end of one main.
-- This photographs a session: the turn executable signals once per settled
-- checkpoint, from inside the store that names the instant, and every one of
-- those frames is written out for the joiner to compare against the pixel
-- claims that were printed beside it.
--
-- The difference from the render harness is only that there are many, and it
-- has two consequences worth stating rather than discovering:
--
--   * **The GPU's registers are checked once, at the first capture.** They
--     cannot change afterwards without the executable having reset the GPU, and
--     decoding a save state eighty-five times to learn the same six facts would
--     be most of the run's wall clock.
--
--   * **The colour-spread guard applies to the opening board and not to every
--     frame.** "At least sixteen distinct colours and no colour over 60%" is
--     the blank-screen guard. It is exactly right for a board and exactly
--     wrong for the information sheet, which is deliberately a screen of
--     paper with letters on it. So the guard is asserted where a
--     board is guaranteed to be (the first checkpoint, which is the board as
--     the content authored it) and every frame is still required to be a
--     representable 15-bit picture with more than one colour in it. A run whose
--     display went blank half way through fails on the second of those.
--
-- The per-cell comparison is not done here, for the same reason it is not done
-- in the render harness: this script cannot see the executable's stdout, and a
-- harness that decided the gate from inside the emulator would be checking one
-- channel. `playstation_turn_probe.c` reads both.
--
-- ---------------------------------------------------------------------------
-- The second camera
--
-- Off unless `GRANDLEON_PLAYSTATION_FILM_DIR` is set, and then it writes every
-- composited frame between two checkpoint ordinals rather than one frame per
-- checkpoint. `film_from`/`film_to` are checkpoint ordinals, inclusive at both
-- ends, and a window is what makes a film a few seconds rather than the whole
-- session.
--
-- It touches nothing the verdict is computed from. `checks`, `failures` and
-- `captures` are not incremented here, no frame written here goes to the
-- joiner's directory, and the checkpoint camera above is unchanged, so a
-- filmed run makes exactly the checks an unfilmed one makes and reaches exactly
-- the same verdict. That is the property that lets a film be evidence: it is
-- only ever a recording of a run that passed, taken by the harness that judged
-- it, in the process that judged it.

-- The three slots this harness answers on, and they are written down in exactly
-- two places: here and in `platform/playstation/src/turn_exe.cpp`. None of them
-- is the render harness's 13, on purpose: a harness written for one executable
-- must not silently photograph the other.
--
-- A board and a campaign screen are photographed on *different* slots because
-- they are different pictures and the strongest thing that can be said about
-- one is false of the other. A board is terrain, tokens, washes and a cursor,
-- so a board holding three colours is a renderer that did not run; a campaign
-- screen is a page of text, so three colours is exactly what a title screen
-- has. Telling them apart here is what lets each be held to its own claim
-- rather than both to the weaker one.
local SLOT = 14
local SCREEN_SLOT = 12
local DONE_SLOT = 15

local band, rshift = bit.band, bit.rshift

local frame_dir = os.getenv('GRANDLEON_PLAYSTATION_FRAME_DIR') or '/out'

-- What the renderer asked the GPU for. `psx_gpu.cpp`'s numbers, repeated here
-- rather than derived from the frame precisely so that a renderer which changed
-- one and produced a consistent-but-different picture is caught rather than
-- followed.
local WANT_WIDTH = 320
local WANT_HEIGHT = 240
local WANT_START_X = 0
local WANT_START_Y = 0
local WANT_RANGE_X0 = 0x200
local WANT_RANGE_X1 = 0x200 + WANT_WIDTH * 8
local WANT_RANGE_Y0 = 16
local WANT_RANGE_Y1 = 16 + WANT_HEIGHT

local checks = 0
local failures = 0
local captures = 0

-- The second camera's settings, and its own counters, kept apart from the ones
-- above on purpose: nothing here may reach the verdict.
local film_dir = os.getenv('GRANDLEON_PLAYSTATION_FILM_DIR')
local film_from = tonumber(os.getenv('GRANDLEON_PLAYSTATION_FILM_FROM') or '') or 1
local film_to = tonumber(os.getenv('GRANDLEON_PLAYSTATION_FILM_TO') or '') or 0
-- A ceiling rather than a budget, and it is above anything a real window
-- reaches: the whole eighty-five-checkpoint session is 2,270 retraces, which is
-- 349 MB of raw frames, and this stops at 4,000. It exists so that a harness
-- pointed at a longer-running executable cannot quietly fill a disk; the run
-- script fails the run rather than shipping a truncated film.
local film_limit = tonumber(os.getenv('GRANDLEON_PLAYSTATION_FILM_LIMIT') or '') or 4000
local film_written = 0
local film_truncated = false

local function say(text)
  print('HARNESS ' .. text)
end

local function expect(condition, name)
  checks = checks + 1
  if not condition then failures = failures + 1 end
  say('CHECK ' .. name .. (condition and ' PASS' or ' FAIL'))
end

local function expect_equal(actual, wanted, name)
  expect(actual == wanted,
         string.format('%s (%s, wanted %s)', name, tostring(actual),
                       tostring(wanted)))
end

-- The GPU registers, out of a save state. GP1 is write-only on the hardware and
-- PCSX-Redux routes the port straight into the GPU without mirroring it, so the
-- emulator's own 256-entry shadow is the only legible copy.
local function gpu_registers()
  local protoc = require('protoc')
  local pb = require('pb')
  local compiler = protoc.new()
  assert(compiler:load(PCSX.getSaveStateProtoSchema()))
  local state = pb.decode('SaveState', tostring(PCSX.createSaveState()))
  local control = state.gpu.control
  local function gp1(command)
    local base = command * 4
    return control:byte(base + 1) + control:byte(base + 2) * 256
         + control:byte(base + 3) * 65536 + control:byte(base + 4) * 16777216
  end
  return state.gpu.status, gp1
end

local function check_registers()
  local decoded, status, gp1 = pcall(gpu_registers)
  expect(decoded, 'the GPU registers decode')
  if not decoded then
    say('DETAIL ' .. tostring(status))
    return
  end

  say(string.format('GPUSTAT %08x', status))
  expect(band(rshift(status, 23), 1) == 0,
         'GPUSTAT says the display is enabled')
  expect_equal(band(gp1(0x03), 1), 0, 'GP1(03) display enable')

  local start = gp1(0x05)
  expect_equal(band(start, 0x3FF), WANT_START_X, 'GP1(05) display start x')
  expect_equal(band(rshift(start, 10), 0x1FF), WANT_START_Y,
               'GP1(05) display start y')

  local horizontal = gp1(0x06)
  expect_equal(band(horizontal, 0xFFF), WANT_RANGE_X0,
               'GP1(06) horizontal range x0')
  expect_equal(band(rshift(horizontal, 12), 0xFFF), WANT_RANGE_X1,
               'GP1(06) horizontal range x1')

  local vertical = gp1(0x07)
  expect_equal(band(vertical, 0x3FF), WANT_RANGE_Y0,
               'GP1(07) vertical range y0')
  expect_equal(band(rshift(vertical, 10), 0x3FF), WANT_RANGE_Y1,
               'GP1(07) vertical range y1')

  expect_equal(band(rshift(status, 17), 3), 1, 'GPUSTAT horizontal resolution')
  expect_equal(band(rshift(status, 19), 1), 0, 'GPUSTAT vertical resolution')
  expect_equal(band(rshift(status, 21), 1), 0,
               'GPUSTAT colour depth is 15-bit')
end

-- Whether the first frame of each kind has been judged. The strong claims below
-- are about the first one the run produces, because that is the one that says
-- the renderer works at all; every frame after it is held to the claims that
-- are true of all of them, plus the joiner's exact pixels.
local board_judged = false
local screen_judged = false

local function measure(is_board)
  local index = captures
  captures = captures + 1
  if index == 0 then check_registers() end

  local shot = PCSX.GPU.takeScreenShot()
  local width = tonumber(shot.width)
  local height = tonumber(shot.height)
  if index == 0 then
    expect_equal(width, WANT_WIDTH, 'the frame is as wide as the display')
    expect_equal(height, WANT_HEIGHT, 'the frame is as tall as the display')
    expect_equal(shot.data.size, width * height * 2,
                 'the frame is 15-bit colour')
  elseif width ~= WANT_WIDTH or height ~= WANT_HEIGHT then
    expect(false, string.format('checkpoint %d keeps the display size', index))
  end

  local pixels = width * height
  local histogram = {}
  local distinct = 0
  local largest = 0
  local representable = true
  for offset = 0, shot.data.size - 2, 2 do
    local colour = shot.data[offset] + shot.data[offset + 1] * 256
    -- Every colour this executable can put on screen comes out of a CLUT this
    -- repository generated or an interface constant it chose, and neither ever
    -- sets the semi-transparency bit.
    if band(colour, 0x8000) ~= 0 then representable = false end
    local seen = (histogram[colour] or 0) + 1
    if seen == 1 then distinct = distinct + 1 end
    histogram[colour] = seen
    if seen > largest then largest = seen end
  end

  -- Every frame: a picture, in colours this art can make, with something on it.
  expect(representable,
         string.format('checkpoint %d is drawn in colours this art can produce',
                       index))
  expect(distinct > 1,
         string.format('checkpoint %d is not a single flat colour', index))
  -- The first board, and only it: a sheet is deliberately a screen of paper.
  if is_board and not board_judged then
    board_judged = true
    expect(distinct >= 16,
           'the opening board holds at least sixteen distinct colours')
    expect(largest * 100 < pixels * 60,
           'no single colour covers more than 60% of the opening board')
  end
  -- The first campaign screen, and only it. A page is paper, whatever the panel
  -- stands on, and ink, so three colours is the floor and two is a panel with
  -- nothing written on it. The second claim is what says the third colour is
  -- *words*: a title screen's ink covers 843 of 76,800 pixels, and a page that
  -- drew one letter would cover about fifteen.
  --
  -- These are deliberately looser than the board's pair and they are not the
  -- weight this frame carries. A screen's real proof is in the joiner, which
  -- compares the panel at both ends, the first ink pixel of the first written
  -- row, the caret, the footer and the scene's own backdrop band against the
  -- executable's claims, the GPU's readback and this frame: five exact pixels
  -- with no tolerance, which no histogram can be.
  if not is_board and not screen_judged then
    screen_judged = true
    local commonest, second = 0, 0
    for _, seen in pairs(histogram) do
      if seen > commonest then
        second = commonest
        commonest = seen
      elseif seen > second then
        second = seen
      end
    end
    expect(distinct >= 3, 'the opening screen is paper, ground and ink')
    expect(pixels - commonest - second >= 256,
           'and at least 256 pixels of it are words')
  end

  local path = string.format('%s/frame-%04d.bin', frame_dir, index)
  local out = Support.File.open(path, 'TRUNCATE')
  out:writeMoveSlice(shot.data)
  out:close()
  say(string.format('FRAME %d %dx%d distinct %d largest %d%%', index, width,
                    height, distinct,
                    pixels > 0 and math.floor(largest * 100 / pixels) or 0))
  -- Where this checkpoint fell in the film, so that choosing a window is
  -- reading rather than guessing.
  if film_dir then
    say(string.format('FILM-AT %d checkpoint %d', film_written, index + 1))
  end
end

local function capture_kind(is_board)
  local ok, err = pcall(measure, is_board)
  if not ok then
    say('CHECK the capture completes FAIL')
    say('DETAIL ' .. tostring(err))
    say(string.format('RESULT FAIL %d/%d', checks - failures - 1, checks + 1))
    PCSX.quit(1)
  end
  if failures ~= 0 then
    -- End the run here rather than let the executable write a passing exit code
    -- over the top of a failed measurement.
    say(string.format('FRAMES %d %dx%d', captures, WANT_WIDTH, WANT_HEIGHT))
    say(string.format('RESULT FAIL %d/%d', checks - failures, checks))
    PCSX.quit(1)
  end
end

local function capture() capture_kind(true) end
local function capture_screen() capture_kind(false) end

PCSX.execSlots[SLOT] = capture
PCSX.execSlots[SCREEN_SLOT] = capture_screen

-- The film, one frame per vertical retrace.
--
-- A retrace is where a single-buffered machine's picture is finished: the
-- executable draws and then waits for one, so what the display holds at this
-- instant is a whole frame the console showed. This listens to the emulator's
-- own retrace event, which is why the film needs no wall clock to say how long
-- a frame lasted.
local function film()
  if captures < film_from then return end
  if film_to > 0 and captures > film_to then return end
  if film_written >= film_limit then
    film_truncated = true
    return
  end
  local shot = PCSX.GPU.takeScreenShot()
  local path = string.format('%s/film-%05d.bin', film_dir, film_written)
  local out = Support.File.open(path, 'TRUNCATE')
  out:writeMoveSlice(shot.data)
  out:close()
  film_written = film_written + 1
end

-- Global, and it has to be. `createEventListener` hands back an object that
-- *is* the subscription: let it be collected and the listener goes with it, and
-- a `local` in a `-dofile` chunk nothing else closes over is collectable the
-- moment the chunk returns. `playstation_probe.lua` gets away with dropping its
-- listener only because that one fires on the very first retrace, before a
-- collection can happen; this one has to survive forty-odd checkpoints first,
-- and does not get away with it. Dropped, it wrote no frames at all.
grandleon_film_listener = nil

if film_dir then
  grandleon_film_listener = PCSX.Events.createEventListener('GPU::Vsync', function()
    local ok, err = pcall(film)
    if not ok then
      say('FILM-FAILED ' .. tostring(err))
      PCSX.quit(1)
    end
  end)
  say(string.format('FILM-DIR %s checkpoints %d..%s', film_dir, film_from,
                    film_to > 0 and tostring(film_to) or 'end'))
end

-- The verdict, written when the guest says it has finished settling, because
-- only the guest knows that. A second slot rather than an emulator event: the
-- events this emulator publishes are about the machine's own lifecycle (a
-- shell reached, a state loaded) and none of them means "the program has
-- stopped photographing itself". The executable signals this one immediately
-- before it returns from `main`, and the joiner requires the line to be
-- present, so a run that ended early is a failure rather than a silence.
PCSX.execSlots[DONE_SLOT] = function()
  if film_dir then
    say(string.format('FILM %d frames%s', film_written,
                      film_truncated and ' TRUNCATED' or ''))
  end
  say(string.format('FRAMES %d %dx%d', captures, WANT_WIDTH, WANT_HEIGHT))
  say(string.format('RESULT %s %d/%d', failures == 0 and 'PASS' or 'FAIL',
                    checks - failures, checks))
end

say('installed on execSlot ' .. SLOT)
say('FRAME-DIR ' .. frame_dir)

File: platform/playstation/harness/playstation_turn_probe.c
// SPDX-License-Identifier: MIT
/* The half of the turn check that decides the gate.
 *
 * Usage:
 *   playstation_turn_probe <log> <expectations.txt> <frame-dir>
 *
 * It reads three things three different machines produced independently, and
 * requires all of them to agree:
 *
 *   - **the expectations**, derived on the host by `grandleon_playstation_expect`
 *     before the executable was built: the controller script, and the
 *     `CHECKPOINT`/`FACT` transcript the rules say it adds up to;
 *   - **the log**, which is what the R3000A said. The same transcript, plus a
 *     `PROBE` line per claimed pixel (a screen coordinate and the five-bit
 *     colour the executable computed from `grandleon::view`, the art library
 *     and its own interface constants) and a `READBACK` line saying what the
 *     GPU then stored at that address, fetched through GP0(0xC0);
 *   - **the frames**, one per checkpoint, captured by the emulator's Lua
 *     harness from the composited display at the instant the executable named,
 *     and written out raw: 15-bit little-endian halfwords, cropped to exactly
 *     the active display, one halfword per pixel, no padding.
 *
 * ---------------------------------------------------------------------------
 * Why this program knows nothing
 *
 * It never reads the package, never sees the art, and knows nothing about the
 * board or the rules, the same discipline `playstation_probe.c` follows and
 * for the same reason. It links nothing from this repository at all, so it
 * cannot agree with the executable by sharing a mistake with it. Its only
 * numbers of its own are the frame's dimensions, and it takes those from the
 * harness's own `FRAMES` line rather than assuming them.
 *
 * There is no tolerance anywhere. A PlayStation stores five bits per channel
 * and a frame is those bits verbatim.
 *
 * ---------------------------------------------------------------------------
 * What agreement proves
 *
 *   transcript  the console reached the state the rules say those presses reach
 *   claim       the executable's arithmetic over the projection and the CLUT
 *   readback    the GPU's rasteriser and CLUT unit, out of VRAM
 *   frame       the emulator's display window over that same VRAM
 *
 * A client that reaches the right state and draws the wrong picture fails on
 * the second; a client that draws a right picture into the wrong place fails on
 * the third; a client that agrees with itself about a state the engine never
 * reaches fails on the first. The expectations exist before the executable does,
 * so no run can be made to pass by adjusting one.
 */

#include <stdarg.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

enum {
    max_line = 512,
    max_label = 64,
    max_probes_per_block = 512,
    max_expected = 262144,
    max_frames = 512
};

struct claim {
    char label[max_label];
    int x;
    int y;
    int red;
    int green;
    int blue;
    int has_readback;
    int read_red;
    int read_green;
    int read_blue;
};

static struct claim claims[max_probes_per_block];
static int claim_count = 0;
static int declared_probes = -1;

static char *expected_lines[max_expected];
static int expected_count = 0;
static int expected_index = 0;

static int checks = 0;
static int failures = 0;

static int checkpoints_seen = 0;
static int blocks_compared = 0;
static int pixels_compared = 0;
static char block_name[max_label] = "";
static int block_ok = 1;
static int block_open = 0;

static int frame_width = 0;
static int frame_height = 0;
static int frames_reported = -1;
static const char *frame_dir = NULL;
static uint16_t *frame_pixels = NULL;
/* Pixels `frame_pixels` was allocated for, so the read below is bounded by the
 * buffer it writes into rather than by an argument about the rest of the file. */
static size_t frame_capacity = 0;
static int frame_loaded = -1;

static int harness_result_seen = 0;
static int harness_passed = 0;
static int rom_result_seen = 0;
static int rom_passed = 0;

/* The script, counted at both ends.
 *
 * A harness with pad ports to play the `PRESS` lines into could not let the
 * expectation file and the run differ. Here the script is compiled into the
 * executable and this file only records it, so a script edited without
 * re-deriving, or re-derived without rebuilding, is a real way for the two
 * halves to drift. The transcript comparison would catch it, but it
 * would catch it as a wall of mismatched facts; these three numbers say what
 * actually happened. */
static int script_presses = 0;
static int declared_script = -1;
static int played_presses = -1;

static void expect(int condition, const char *format, ...)
    __attribute__((format(printf, 2, 3)));

static void expect(int condition, const char *format, ...) {
    va_list arguments;
    ++checks;
    if (!condition) ++failures;
    fputs(condition ? "ok   " : "FAIL ", stdout);
    va_start(arguments, format);
    vprintf(format, arguments);
    va_end(arguments);
    fputc('\n', stdout);
}

static void report(const char *format, ...) {
    va_list arguments;
    va_start(arguments, format);
    vprintf(format, arguments);
    va_end(arguments);
}

static void trim(char *line) {
    size_t length = strlen(line);
    while (length > 0 && (line[length - 1] == '\n' || line[length - 1] == '\r' ||
                          line[length - 1] == ' ')) {
        line[--length] = '\0';
    }
}

/* ------------------------------------------------------------------------- */
/* The frames                                                                */
/* ------------------------------------------------------------------------- */

static int load_frame(int index) {
    char path[1024];
    FILE *file;
    size_t wanted;
    size_t read;

    if (frame_loaded == index) return 1;
    if (frame_dir == NULL || frame_width <= 0 || frame_height <= 0) return 0;
    wanted = (size_t)frame_width * (size_t)frame_height;
    if (frame_pixels == NULL) {
        frame_pixels = malloc(wanted * sizeof *frame_pixels);
        if (frame_pixels == NULL) return 0;
        frame_capacity = wanted;
    }
    /* The buffer is sized once, from the frame size `main` settled by reading
     * the whole log before a single block was compared, and `handle` refuses a
     * later line that disagrees with it rather than adopting it. This asks the
     * same question again at the only place it decides anything: how many
     * halfwords `fread` is about to write. Without it, a second `HARNESS
     * FRAMES` line naming a larger frame reads the larger frame into the
     * smaller allocation, and the emulator's output chooses how far past the
     * end of the heap block it writes. */
    if (wanted > frame_capacity) return 0;
    snprintf(path, sizeof path, "%s/frame-%04d.bin", frame_dir, index);
    file = fopen(path, "rb");
    if (file == NULL) return 0;
    read = fread(frame_pixels, sizeof *frame_pixels, wanted, file);
    fclose(file);
    if (read != wanted) return 0;
    frame_loaded = index;
    return 1;
}

/* ------------------------------------------------------------------------- */
/* The transcript                                                            */
/* ------------------------------------------------------------------------- */

static void expect_transcript(const char *line) {
    if (expected_index >= expected_count) {
        report("       the console said more than the host derived: %s\n", line);
        block_ok = 0;
        return;
    }
    if (strcmp(line, expected_lines[expected_index]) != 0) {
        report("       host:    %s\n", expected_lines[expected_index]);
        report("       console: %s\n", line);
        block_ok = 0;
    }
    ++expected_index;
}

/* Every claim in the block just closed, against the frame captured for it.
 *
 * Four assertions a checkpoint, with a printed line per disagreement inside
 * them. Aggregating this way rather than asserting per pixel is deliberate: a
 * checkpoint is the unit a person debugs, one wrong pixel fails its whole
 * block, and the detail lines say which pixel it was. */
static void compare_block(void) {
    int index;
    int loaded;
    int readbacks_agree = 1;
    int frame_agrees = 1;

    if (!block_open) return;
    block_open = 0;

    loaded = load_frame(blocks_compared);
    expect(loaded, "checkpoint %s has a captured frame", block_name);

    expect(declared_probes == claim_count,
           "checkpoint %s printed the %d probes it counted (saw %d)",
           block_name, declared_probes, claim_count);

    for (index = 0; index < claim_count && loaded; ++index) {
        const struct claim *entry = &claims[index];
        uint16_t value;
        int red;
        int green;
        int blue;

        if (entry->x < 0 || entry->y < 0 || entry->x >= frame_width ||
            entry->y >= frame_height) {
            report("       %s/%s is outside the frame at (%d, %d)\n",
                   block_name, entry->label, entry->x, entry->y);
            frame_agrees = 0;
            continue;
        }
        if (!entry->has_readback) {
            report("       %s/%s has no VRAM readback\n", block_name,
                   entry->label);
            readbacks_agree = 0;
            continue;
        }
        if (entry->read_red != entry->red || entry->read_green != entry->green ||
            entry->read_blue != entry->blue) {
            report("       %s/%s claimed %d %d %d, the GPU stored %d %d %d\n",
                   block_name, entry->label, entry->red, entry->green,
                   entry->blue, entry->read_red, entry->read_green,
                   entry->read_blue);
            readbacks_agree = 0;
            continue;
        }
        value = frame_pixels[(size_t)entry->y * (size_t)frame_width +
                             (size_t)entry->x];
        red = value & 0x1F;
        green = (value >> 5) & 0x1F;
        blue = (value >> 10) & 0x1F;
        ++pixels_compared;
        if (red != entry->red || green != entry->green || blue != entry->blue) {
            report("       %s/%s at (%d, %d) claimed %d %d %d, the display "
                   "shows %d %d %d\n",
                   block_name, entry->label, entry->x, entry->y, entry->red,
                   entry->green, entry->blue, red, green, blue);
            frame_agrees = 0;
        }
    }

    expect(readbacks_agree, "checkpoint %s stored every colour it claimed",
           block_name);
    expect(frame_agrees, "checkpoint %s displays every colour it claimed",
           block_name);
    expect(block_ok, "checkpoint %s matches the host's derivation", block_name);
    ++blocks_compared;
    claim_count = 0;
    declared_probes = -1;
    block_ok = 1;
}

static void open_block(const char *line) {
    compare_block();
    ++checkpoints_seen;
    snprintf(block_name, sizeof block_name, "%.*s", (int)sizeof block_name - 1,
             line + 11);
    block_open = 1;
    block_ok = 1;
    claim_count = 0;
    declared_probes = -1;
    expect_transcript(line);
}

static struct claim *find_claim(const char *label) {
    int index;
    for (index = claim_count - 1; index >= 0; --index) {
        if (strcmp(claims[index].label, label) == 0) return &claims[index];
    }
    return NULL;
}

static void handle(const char *line) {
    char label[max_label];
    int x;
    int y;
    int red;
    int green;
    int blue;
    int count;

    if (strncmp(line, "CHECKPOINT ", 11) == 0) {
        open_block(line);
        return;
    }
    if (strncmp(line, "FACT ", 5) == 0) {
        expect_transcript(line);
        return;
    }
    if (sscanf(line, "PROBE %63s %d %d %d %d %d", label, &x, &y, &red, &green,
               &blue) == 6) {
        struct claim *entry;
        if (claim_count >= max_probes_per_block) {
            expect(0, "checkpoint %s stays inside the probe capacity",
                   block_name);
            return;
        }
        if (find_claim(label) != NULL) {
            expect(0, "checkpoint %s names %s once", block_name, label);
        }
        entry = &claims[claim_count++];
        memset(entry, 0, sizeof *entry);
        snprintf(entry->label, sizeof entry->label, "%s", label);
        entry->x = x;
        entry->y = y;
        entry->red = red;
        entry->green = green;
        entry->blue = blue;
        return;
    }
    if (sscanf(line, "READBACK %63s %d %d %d", label, &red, &green, &blue) == 4) {
        struct claim *entry = find_claim(label);
        if (entry == NULL) {
            expect(0, "checkpoint %s reads back only what it claimed (%s)",
                   block_name, label);
            return;
        }
        entry->has_readback = 1;
        entry->read_red = red;
        entry->read_green = green;
        entry->read_blue = blue;
        return;
    }
    if (sscanf(line, "PROBE-COUNT %d", &count) == 1) {
        declared_probes = count;
        return;
    }
    if (sscanf(line, "HARNESS FRAMES %d %dx%d", &count, &x, &y) == 3) {
        frames_reported = count;
        /* The size is not taken from here. `main` settled it by reading the
         * whole log first, precisely because the frame buffer has to be
         * allocated before the first block is compared and this line arrives
         * after it. What is left to do is disagree: a second line naming a
         * different size is the harness contradicting itself, and a probe
         * whose entire job is to refuse disagreement cannot quietly adopt the
         * newer number over a buffer already sized for the older one.
         *
         * Reported only when it happens, so a run of the one line a real
         * harness prints adds no check to the tally. */
        if (x != frame_width || y != frame_height) {
            expect(0, "the harness reports one frame size (%dx%d, then %dx%d)",
                   frame_width, frame_height, x, y);
        }
        return;
    }
    if (strncmp(line, "LAYOUT ", 7) == 0) {
        const char* at = strstr(line, " script ");
        if (at != NULL && sscanf(at, " script %d", &count) == 1) {
            declared_script = count;
        }
        return;
    }
    if (strncmp(line, "SESSION ", 8) == 0) {
        const char* at = strstr(line, " presses ");
        if (at != NULL && sscanf(at, " presses %d", &count) == 1) {
            played_presses = count;
        }
        return;
    }
    if (strncmp(line, "HARNESS RESULT ", 15) == 0) {
        harness_result_seen = 1;
        harness_passed = strncmp(line + 15, "PASS", 4) == 0;
        return;
    }
    if (strncmp(line, "RESULT ", 7) == 0) {
        rom_result_seen = 1;
        rom_passed = strncmp(line + 7, "PASS", 4) == 0;
        return;
    }
}

static int read_expectations(const char *path) {
    FILE *file = fopen(path, "r");
    char line[max_line];
    int in_transcript = 0;

    if (file == NULL) return 0;
    while (fgets(line, sizeof line, file) != NULL) {
        trim(line);
        if (line[0] == '#' || line[0] == '\0') continue;
        if (strcmp(line, "SCRIPT") == 0) continue;
        if (strcmp(line, "TRANSCRIPT") == 0) {
            in_transcript = 1;
            continue;
        }
        if (!in_transcript) {
            if (strncmp(line, "PRESS ", 6) == 0) ++script_presses;
            continue;
        }
        if (expected_count >= max_expected) {
            fclose(file);
            return 0;
        }
        expected_lines[expected_count] = malloc(strlen(line) + 1);
        if (expected_lines[expected_count] == NULL) {
            fclose(file);
            return 0;
        }
        strcpy(expected_lines[expected_count], line);
        ++expected_count;
    }
    fclose(file);
    return 1;
}

int main(int argc, char **argv) {
    FILE *log;
    char line[max_line];

    if (argc != 4) {
        fprintf(stderr,
                "usage: %s <log> <expectations.txt> <frame-dir>\n",
                argc > 0 ? argv[0] : "playstation_turn_probe");
        return 2;
    }
    frame_dir = argv[3];

    if (!read_expectations(argv[2])) {
        fprintf(stderr, "cannot read the expectations at %s\n", argv[2]);
        return 2;
    }

    log = fopen(argv[1], "r");
    if (log == NULL) {
        fprintf(stderr, "cannot read %s\n", argv[1]);
        return 2;
    }
    /* The frame size has to be known before the first block is compared, and
     * the harness prints it at the end, so the log is read once for the
     * `HARNESS FRAMES` line and once for everything else. It is a few hundred
     * kilobytes and this is a great deal simpler than buffering the blocks. */
    while (fgets(line, sizeof line, log) != NULL) {
        int count;
        int x;
        int y;
        trim(line);
        if (sscanf(line, "HARNESS FRAMES %d %dx%d", &count, &x, &y) == 3) {
            frames_reported = count;
            frame_width = x;
            frame_height = y;
        }
    }
    expect(frame_width > 0 && frame_height > 0,
           "the harness reported a frame size (%dx%d)", frame_width,
           frame_height);
    if (frame_width <= 0 || frame_height <= 0) {
        fclose(log);
        report("RESULT FAIL %d/%d\n", checks - failures, checks);
        return 1;
    }

    rewind(log);
    while (fgets(line, sizeof line, log) != NULL) {
        trim(line);
        handle(line);
    }
    fclose(log);
    compare_block();

    /* Report integrity, before anything is concluded from it. */
    expect(checkpoints_seen > 0, "the console settled somewhere");
    expect(expected_count > 0, "the host derived a transcript");
    expect(script_presses > 0, "the host derived a script (%d presses)",
           script_presses);
    expect(declared_script == script_presses,
           "the executable carries the script the host derived (%d, derived %d)",
           declared_script, script_presses);
    expect(played_presses == script_presses,
           "the executable played every press of it (%d of %d)", played_presses,
           script_presses);
    expect(expected_index == expected_count,
           "every line the host derived was reported (%d of %d)", expected_index,
           expected_count);
    expect(frames_reported == checkpoints_seen,
           "the harness captured one frame per checkpoint (%d frames, %d "
           "checkpoints)",
           frames_reported, checkpoints_seen);
    expect(pixels_compared > 0, "at least one pixel was compared");
    expect(harness_result_seen, "the emulator's harness reached a verdict");
    expect(harness_passed, "the emulator's harness passed");
    expect(rom_result_seen, "the executable reached a verdict");
    expect(rom_passed, "the executable passed");

    report("CHECKPOINTS %d  BLOCKS %d  PIXELS %d\n", checkpoints_seen,
           blocks_compared, pixels_compared);
    report("RESULT %s %d/%d\n", failures == 0 ? "PASS" : "FAIL",
           checks - failures, checks);
    return failures == 0 ? 0 : 1;
}
